nixos/modules/contracts/default.nix

```
@@ -0,0 +1,154 @@
+{ lib, ... }:
+let
+  inherit (lib) mkOption;
+  inherit (lib.types) attrs attrsOf functionTo submodule listOf str deferredModule optionType;
+in
+{
+  options.contracts = mkOption {
+    description = ''
+      Base option for a contract.
+
+      To create a new contract, add an instance of `config.contracts.<name>`
+      and define the `meta`, `input` and `output` options.
+      The `consumer` and `provider` options will then be setup automatically
+      and contain respectively the type of a consumer and provider
+      of this new contract.
+
+      To use the `<name>` contract, create an option with either the
+      `config.contracts.<name>.consumer` or `config.contracts.<name>.provider`
+      type.
+
+      The `behaviorTest` option is used to ensure all `provider` of a contract
+      behave the same way.
+    '';
+    type =
+      attrsOf (
```

```suggestion
      lazyAttrsOf (
```

This is usually preferable for attrsets that are consumed primarily through attribute selection.
It allows the evaluation of more self-references than `attrsOf` would.

https://nixos.org/manual/nixos/unstable/#sec-option-types-composed
(which ultimately boils down to [this](https://github.com/NixOS/nixpkgs/pull/428330/files) non-module trade-off, after seeing through `mkIf` which ultimately isn't magic)